pkg/updater/updater.go

```diff
@@ -131,97 +117,185 @@ func lockGroup(ctx context.Context, client gcs.ConditionalClient, path gcs.Path,
-	defer growMaxUpdateArea()
+// Updates test groups with the specified freq.
+//
+// Filters down to a single group when set.
+// Returns after all groups updated once if freq is zero.
+func Update(parent context.Context, client gcs.ConditionalClient, mets *Metrics, configPath gcs.Path, gridPrefix string, groupConcurrency int, group string, updateGroup GroupUpdater, write bool, freq time.Duration) error {
 	ctx, cancel := context.WithCancel(parent)
 	defer cancel()
 	log := logrus.WithField("config", configPath)
-	cfg, err := config.ReadGCS(ctx, client, configPath)
+
+	var q config.TestGroupQueue
+
+	gen, generations, err := updateTestGroups(ctx, client, &q, configPath, gridPrefix, group, freq)
 	if err != nil {
-		return fmt.Errorf("read config: %w", err)
+		return err
 	}
-	log.WithField("groups", len(cfg.TestGroups)).Info("Updating test groups")
-
-	groups := make(chan configpb.TestGroup)
+	var lock sync.Mutex
 	var wg sync.WaitGroup
+	wg.Add(groupConcurrency)
 	defer wg.Wait()
-	defer close(groups)
-
-	var generations map[string]int64
-
+	channel := make(chan *configpb.TestGroup) // TODO(fejta): pass into this function
+	defer close(channel)
 	for i := 0; i < groupConcurrency; i++ {
-		wg.Add(1)
 		go func() {
-			for tg := range groups {
-				if err := update(ctx, client, log, &tg, configPath, gridPrefix, updateGroup, write, generations); err != nil {
+			defer wg.Done()
+			for tg := range channel {
+				log := log.WithField("group", tg.Name)
+				tgp, err := testGroupPath(configPath, gridPrefix, tg.Name)
+				if err != nil {
+					log.WithError(err).Error("Bad path")
 					mets.Error()
-				} else {
-					mets.Success()
+					continue
+				}
+				gen, ok := generations[tg.Name]
+				if !ok {
+					gen = -1
+				}
+				if err := update(ctx, client, log, tg, *tgp, updateGroup, write, gen); err != nil {
+					log.WithError(err).Error("Error updating group")
+					mets.Error()
+					continue
+				}
+				mets.Success()
+				growMaxUpdateArea()
+				if attrs, err := client.Stat(ctx, *tgp); err == nil {
+					lock.Lock()
+					generations[tg.Name] = attrs.Generation
+					lock.Unlock()
 				}
 			}
-			wg.Done()
 		}()
 	}
 
-	if group != "" { // Just a specific group
-		tg := config.FindTestGroup(group, cfg)
-		if tg == nil {
-			return errors.New("group not found")
-		}
-		groups <- *tg
-	} else { // All groups
-		generations, err = sortGroups(ctx, log, client, configPath, gridPrefix, cfg.TestGroups)
-		if err != nil {
-			log.WithError(err).Warning("Failed to sort groups")
+	go func() {
+		cond := storage.Conditions{GenerationNotMatch: gen}
+		client := client.If(&cond, nil)
+		ticker := time.NewTicker(time.Minute)
+		for {
+			now := time.Now()
+			depth, next, when := q.Status()
+			log := log.WithField("depth", depth)
+			if next != nil {
+				log = log.WithField("next", next.Name)
+			}
+			if delay := now.Sub(when).Round(time.Second); delay > 0 {
+				log = log.WithField("delay", delay)
```

Comment:
Delay seems like it might be a useful metric to track instead of cycle duration. What do you think?

Reply:
Done